cmd/query/app/querysvc/query_service.go

```diff
@@ -0,0 +1,121 @@
+)
+
+// QueryService contains span utils required by the query-service.
+type QueryService struct {
+	spanReader       spanstore.Reader
+	dependencyReader dependencystore.Reader
+	options          QueryServiceOptions
+}
+
+// NewQueryService returns a new QueryService.
+func NewQueryService(spanReader spanstore.Reader, dependencyReader dependencystore.Reader, options QueryServiceOptions) *QueryService {
+	qsvc := &QueryService{
+		spanReader:       spanReader,
+		dependencyReader: dependencyReader,
+		options:          options,
+	}
+
+	if qsvc.options.adjuster == nil {
+		qsvc.options.adjuster = adjuster.Sequence(StandardAdjusters...)
+	}
+	return qsvc
+}
+
+// Implement the spanstore.Reader interface
+
+// GetTrace is the queryService implementation of spanstore.Reader.GetTrace
+func (qs QueryService) GetTrace(ctx context.Context, traceID model.TraceID) (*model.Trace, error) {
+	trace, err := qs.spanReader.GetTrace(ctx, traceID)
+	if err == spanstore.ErrTraceNotFound {
+		if qs.options.archiveSpanReader == nil {
+			return nil, err
+		}
+		trace, err = qs.options.archiveSpanReader.GetTrace(ctx, traceID)
+	}
+	return trace, err
+}
+
+// GetServices is the queryService implementation of spanstore.Reader.GetServices
+func (qs QueryService) GetServices(ctx context.Context) ([]string, error) {
+	return qs.spanReader.GetServices(ctx)
+}
+
+// GetOperations is the queryService implementation of spanstore.Reader.GetOperations
+func (qs QueryService) GetOperations(ctx context.Context, service string) ([]string, error) {
+	return qs.spanReader.GetOperations(ctx, service)
+}
+
+// FindTraces is the queryService implementation of spanstore.Reader.FindTraces
+func (qs QueryService) FindTraces(ctx context.Context, query *spanstore.TraceQueryParameters) ([]*model.Trace, error) {
+	return qs.spanReader.FindTraces(ctx, query)
+}
+
+// FindTraceIDs is the queryService implementation of spanstore.Reader.FindTraceIDs
+func (qs QueryService) FindTraceIDs(ctx context.Context, query *spanstore.TraceQueryParameters) ([]model.TraceID, error) {
+	return qs.spanReader.FindTraceIDs(ctx, query)
+}
+
+// ArchiveTrace is the queryService utility to archive traces.
+func (qs QueryService) ArchiveTrace(ctx context.Context, traceID *model.TraceID) error {
+	if qs.options.archiveSpanReader == nil {
+		return errNoArchiveSpanStorage
+	}
+	trace, err := qs.GetTrace(ctx, *traceID)
+	if err != nil {
+		return err
+	}
+
+	var writeErrors []error
+	for _, span := range trace.Spans {
+		err := qs.options.archiveSpanWriter.WriteSpan(span)
+		if err != nil {
+			writeErrors = append(writeErrors, err)
+		}
+	}
+	multierr := multierror.Wrap(writeErrors)
+	return multierr
+}
+
+// Implement the adjuster.Adjuster interface
+
+// Adjust implements adjuster.Adjuster.Adjust
+func (qs QueryService) Adjust(trace *model.Trace) (*model.Trace, error) {
+	return qs.options.adjuster.Adjust(trace)
+}
+
+// Implement the dependencystore.Reader interface
```

Comment:
Initially I thought to implement the reader/writer interface so that the `QueryService` object could be passed around as a `spanstore.Reader` / `spanstore.Writer` object in the APIHandler. But I realise that we're trying to abstract the details of the reader/writer away from the APIHandler, so should I change the functions to pass a QueryService object instead?

Reply:
APIHandler already has a reference to QueryService, so you don't need to pass anything, right?

If there's no strong reason to have QueryService implementing Reader/Writer, I would avoid mentioning it, so that in the future these can become de-coupled. E.g. we may need some extra argument to the QueryService, but not to  the storage APIs.

Reply:
You're right. I've made all the changes. 

Just one last thing - As part of the `handler_test` we have a mock implementation of the reader/writer interfaces [here](https://github.com/jaegertracing/jaeger/blob/master/cmd/query/app/handler_test.go#L136). This needs to be replaced with a mock QueryService implementation?

Reply:
why not use the main QueryService?  In many cases it just delegates to the underlying reader/writer anyway, mocking it doesn't seem that useful.